core/WASM_core.md

```diff
@@ -0,0 +1,350 @@
+
+It has an Apache 2.0 license.
+
+It's owned and maintained by the bytecode alliance, which also owns the WASM/WASI spec. It's backed
+by Mozilla.
+
+### Oh no CGo (and JNI)
+There's a big downside to both Wasmer and Wasmtime. Both of their Go implementations use CGo, and
+Wasmer's Java implementation uses JNI. This blows one of the bigger pluses out of the water.
+Luckily, [wazero](https://github.com/tetratelabs/wazero) exists, which we can use to get a pure Go
+environment. It supports WASI as well, but seemingly [not what we would need for network
+calls](https://wazero.io/specs/).
+
+I can't seem to find anything reasonable for Java. Maybe if we're blessed, such a thing
+will exist by the time we get around to porting Java, but that's a huge if.
+
+## Implementation Strategy
+
+### How much, if any WASI to use?
+
+The more we can lean on WASI, the less we have to expand the Core<-->Lang interface. That's a good
+thing in terms of both immediate work and long-term maintenance. There are some problems though:
+
+* Assuming we need to use `wazero` for non-CGo support, it doesn't appear to be able to handle
+  network calls. We'll test this, though.
+* Calling back into lang gives us opportunities for more user customization, which can be nice.
+  Especially when it comes to the metrics and tracing stuff. gRPC interceptors are also a potential
+  plus.
+
+My inclination for now is to at least *investigate* how much we can get out of WASI. Particularly
+with respect to running from Go.
+
+Unfortunately, my assumption is that we will need to callback into Lang for all the usecases
+mentioned at the top of the doc. The only sys interface things it appears we'll be able to reliably
+use are clocks (which is good, that does come up) and possibly file I/O, which currently does not
+happen in Core (and probably never should) except for in the test server usecase.
+
+
+### Core API changes
+Operating under the assumption that all the major syscalls will need to be handled through callbacks
+into lang. In WASM, calling functions in the host is supported via `import`ing those host functions.
+Ultimately these functions may be bundled together into a trait or something on the Core side, but
+this doc will focus on the individual functions.
+
+Here are the functions I think we'll need to add to the Core<-->Lang interface, beyond the existing
+ones that lang already calls. They're defined as they would appear in Rust, after being imported.
+This is definitely a rough draft, as it will be influenced by runtime selection and whatever
+practical problems we run into along the way.
+
+```rust
+// In reality, everything in these functions is going to need to be serialized and shuttled back
+// and forth as bytes. That makes for uninteresting signatures, though, so instead I show what
+// those signatures would look like if we were using deserialized types directly. The exact details
+// of how that serialization happens are intentionally not included here.
```

Comment:
There is a lot of work in this area for calling into and out of the wasm runtime, but I guess non of it is mature enough?

Reply:
Yeah, honestly there's so much stuff going on and it's so new it's hard to even keep track of what's implemented / just an idea / etc.

Like there's this new proposal here https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md which apparently supersedes interface types, which was the old thing.

I think, at the end of the day, we're going to have to do stuff the manual way until something mature (and by mature, I mean even implemented at all) emerges from the soup.